File: src/utils/dependencyGeometry.ts
import type { Vector2D } from '../types/spatial'

const PORT_STANDOFF = 8
const PORT_INSET = 24

export interface DependencyEndpointGeometry {
  center: Vector2D
  halfW: number
  halfH: number
}

export function dependencyStatusLabel(dependentTitle: string, prerequisiteTitles: string[]): string {
  return `${dependentTitle} waiting on ${prerequisiteTitles.filter(Boolean).join(', ') || 'dependency'}`
}

function clamp(value: number, minimum: number, maximum: number): number {
  return Math.min(maximum, Math.max(minimum, value))
}

/**
 * Dependencies are directional, unlike family-style relations: a
 * prerequisite always leaves the right rail and enters the dependent on its
 * left rail. The y coordinate still slides toward the other endpoint so the
 * route stays short and does not pile every line onto the card center.
 */
export function dependencyAnchors(
  prerequisite: DependencyEndpointGeometry,
  dependent: DependencyEndpointGeometry,
): { start: Vector2D; end: Vector2D } {
  const prerequisiteInset = Math.min(PORT_INSET, Math.max(0, prerequisite.halfH - 1))
  const dependentInset = Math.min(PORT_INSET, Math.max(0, dependent.halfH - 1))
  return {
    start: {
      x: prerequisite.center.x + prerequisite.halfW + PORT_STANDOFF,
      y: clamp(
        dependent.center.y,
        prerequisite.center.y - prerequisite.halfH + prerequisiteInset,
        prerequisite.center.y + prerequisite.halfH - prerequisiteInset,
      ),
    },
    end: {
      x: dependent.center.x - dependent.halfW - PORT_STANDOFF,
      y: clamp(
        prerequisite.center.y,
        dependent.center.y - dependent.halfH + dependentInset,
        dependent.center.y + dependent.halfH - dependentInset,
      ),
    },
  }
}

File: apple/Sources/GrovepadCanvas/Edges/DependencyGeometry.swift
import Foundation
import GrovepadCore

// ---------------------------------------------------------------------------
// Dependency line endpoints (`utils/dependencyGeometry.ts`).
// ---------------------------------------------------------------------------

public enum DependencyGeometry {
    /// Every line on the board keeps the same 0.3-cell gap from the card it
    /// touches, dependencies included.
    public static let portStandoff = EdgeRouting.lineStandoff
    public static let portInset = 24.0

    public static func statusLabel(dependentTitle: String, prerequisiteTitles: [String]) -> String {
        let names = prerequisiteTitles.filter { !$0.isEmpty }.joined(separator: ", ")
        return "\(dependentTitle) waiting on \(names.isEmpty ? "dependency" : names)"
    }

    /// Dependencies are directional, unlike family-style relations: a
    /// prerequisite always leaves the right rail and enters the dependent on
    /// its left rail. The y coordinate still slides toward the other
    /// endpoint so the route stays short and does not pile every line onto
    /// the card center.
    public static func anchors(prerequisite: EdgeNode, dependent: EdgeNode) -> (start: Vector2D, end: Vector2D) {
        let prerequisiteInset = min(portInset, max(0, prerequisite.halfH - 1))
        let dependentInset = min(portInset, max(0, dependent.halfH - 1))
        return (
            Vector2D(
                x: prerequisite.center.x + prerequisite.halfW + portStandoff,
                y: clamp(
                    dependent.center.y,
                    prerequisite.center.y - prerequisite.halfH + prerequisiteInset,
                    prerequisite.center.y + prerequisite.halfH - prerequisiteInset
                )
            ),
            Vector2D(
                x: dependent.center.x - dependent.halfW - portStandoff,
                y: clamp(
                    prerequisite.center.y,
                    dependent.center.y - dependent.halfH + dependentInset,
                    dependent.center.y + dependent.halfH - dependentInset
                )
            )
        )
    }

    /// The full dependency route. Dependencies keep their fixed rails: out of
    /// the prerequisite's right edge, into the dependent's left one, drawn
    /// with the shared edge cubic (`edgePath(start, 'right', end, 'left')`).
    public static func route(prerequisite: EdgeNode, dependent: EdgeNode) -> EdgeRoute {
        let (start, end) = anchors(prerequisite: prerequisite, dependent: dependent)
        return EdgeRoute(
            curve: EdgeRouting.edgeCurve(start: start, startSide: .right, end: end, endSide: .left),
            mid: EdgeRouting.edgeMidpoint(start: start, startSide: .right, end: end, endSide: .left),
            start: start,
            end: end,
            startSide: .right,
            endSide: .left
        )
    }

    /// The web shows the status chip only on a line at least this long.
    public static let statusChipMinLength = 72.0
}
